File: app/src/main/java/com/lambdasoup/watchlater/ui/LauncherActivity.java
/*
 * Copyright (c) 2015 - 2019
 *
 * Maximilian Hille <mh@lambdasoup.com>
 * Juliane Lehmann <jl@lambdasoup.com>
 *
 * This file is part of Watch Later.
 *
 * Watch Later is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Watch Later is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Watch Later.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.lambdasoup.watchlater.ui;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;

import com.lambdasoup.watchlater.BuildConfig;
import com.lambdasoup.watchlater.R;
import com.lambdasoup.watchlater.data.IntentResolverRepository;
import com.lambdasoup.watchlater.viewmodel.LauncherViewModel;

import static android.net.Uri.parse;

public class LauncherActivity extends WatchLaterActivity {

	private static final Uri    EXAMPLE_URI             = parse("https://www.youtube.com/watch?v=dGFSjKuJfrI");
	public static final  Intent EXAMPLE_INTENT          = new Intent(Intent.ACTION_VIEW, EXAMPLE_URI);
	private static final Intent INTENT_YOUTUBE_APP      = new Intent().setData(EXAMPLE_URI).setPackage("com.google.android.youtube");
	private static final Intent INTENT_YOUTUBE_SETTINGS = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.parse("package:com.google.android.youtube"));

	private LauncherViewModel viewModel;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_launcher);

		Toolbar toolbar = findViewById(R.id.toolbar);
		setSupportActionBar(toolbar);

		viewModel = getViewModel(LauncherViewModel.class);
		viewModel.getResolverState().observe(this, this::onResolverStateChanged);

		findViewById(R.id.launcher_youtube_button).setOnClickListener(v -> openYoutubeSettings());
		findViewById(R.id.launcher_example_button).setOnClickListener(v -> openExampleVideo());
	}

	private void onResolverStateChanged(IntentResolverRepository.ResolverState resolverState) {
		View view = findViewById(R.id.launcher_youtube_action);
		switch (resolverState) {
			case OK:
				view.setVisibility(View.GONE);
				break;
			case YOUTUBE_ONLY:
				view.setVisibility(View.VISIBLE);
				break;
		}
	}

	@Override
	protected void onResume() {
		super.onResume();

		viewModel.update();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		MenuInflater inflater = getMenuInflater();
		inflater.inflate(R.menu.menu_add, menu);
		return super.onCreateOptionsMenu(menu);
	}

	@Override
	public boolean onOptionsItemSelected(@NonNull MenuItem item) {
		switch (item.getItemId()) {
			case R.id.menu_about:
				startActivity(new Intent(this, AboutActivity.class));
				return true;
			case R.id.menu_help:
				startActivity(new Intent(this, HelpActivity.class));
				return true;
			case R.id.menu_privacy:
				startActivity(new Intent(Intent.ACTION_VIEW, parse("https://lambdasoup.com/privacypolicy-watchlater/")));
				return true;
			case R.id.menu_store:
				startActivity(new Intent(Intent.ACTION_VIEW, parse("https://play.google.com/store/apps/details?id=" + BuildConfig.APPLICATION_ID)));
				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	private void openExampleVideo() {
		startActivity(EXAMPLE_INTENT);
	}

	private void openYoutubeSettings() {
		// check if youtube is installed
		if (getPackageManager().resolveActivity(INTENT_YOUTUBE_APP, PackageManager.MATCH_DEFAULT_ONLY) == null) {
			Toast.makeText(this, R.string.no_youtube_installed, Toast.LENGTH_SHORT).show();
			return;
		}
		if (getPackageManager().resolveActivity(INTENT_YOUTUBE_SETTINGS, PackageManager.MATCH_DEFAULT_ONLY) == null) {
			Toast.makeText(this, R.string.no_youtube_settings_activity, Toast.LENGTH_SHORT).show();
			return;
		}
		startActivity(INTENT_YOUTUBE_SETTINGS);
	}

}

File: app/src/main/java/com/lambdasoup/watchlater/ui/LauncherActivity.kt
/*
 * Copyright (c) 2015 - 2022
 *
 * Maximilian Hille <mh@lambdasoup.com>
 * Juliane Lehmann <jl@lambdasoup.com>
 *
 * This file is part of Watch Later.
 *
 * Watch Later is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Watch Later is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Watch Later.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.lambdasoup.watchlater.ui

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.lambdasoup.watchlater.BuildConfig
import com.lambdasoup.watchlater.R
import com.lambdasoup.watchlater.WatchLaterApplication
import com.lambdasoup.watchlater.viewmodel.LauncherViewModel
import com.lambdasoup.watchlater.viewmodel.LauncherViewModel.Event

class LauncherActivity : AppCompatActivity() {

    private val vm: LauncherViewModel by viewModels {
        (applicationContext as WatchLaterApplication).viewModelProviderFactory
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_launcher)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        vm.model.observe(this, { render(it) })
        vm.events.observe(this, { event ->
            when (event) {
                Event.OpenYouTubeSettings -> openYoutubeSettings()
                Event.OpenExample -> openExampleVideo()
                Event.OpenWatchLaterSettings -> openWatchLaterSettings()
            }
        })
    }

    private fun render(model: LauncherViewModel.Model) {
        findViewById<View>(R.id.launcher_youtube_button).setOnClickListener { vm.onYoutubeSettings() }
        findViewById<View>(R.id.launcher_watchlater_button).setOnClickListener { vm.onWatchLaterSettings() }
        findViewById<View>(R.id.launcher_example_button).setOnClickListener { vm.onTryExample() }

        val watchLaterView = findViewById<View>(R.id.launcher_watchlater_action)
        watchLaterView.visibility = if (model.resolverProblems != null &&
            ((model.resolverProblems.verifiedDomainsMissing > 0) || !model.resolverProblems.watchLaterIsDefault)) {
            View.VISIBLE
        } else {
            View.GONE
        }
    }

    override fun onResume() {
        super.onResume()
        vm.onResume()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_add, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_about -> {
                startActivity(Intent(this, AboutActivity::class.java))
                true
            }
            R.id.menu_help -> {
                startActivity(Intent(this, HelpActivity::class.java))
                true
            }
            R.id.menu_privacy -> {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://lambdasoup.com/privacypolicy-watchlater/")))
                true
            }
            R.id.menu_store -> {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + BuildConfig.APPLICATION_ID)))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun openExampleVideo() {
        startActivity(EXAMPLE_INTENT)
    }

    private fun openYoutubeSettings() {
        val action = when {
            Build.VERSION.SDK_INT >= 31 -> Settings.ACTION_APP_OPEN_BY_DEFAULT_SETTINGS
            else -> Settings.ACTION_APPLICATION_DETAILS_SETTINGS
        }
        val intent = Intent(action, Uri.parse("package:com.google.android.youtube"))
        if (packageManager.resolveActivity(intent, PackageManager.MATCH_DEFAULT_ONLY ) == null) {
            Toast.makeText(this, R.string.no_youtube_settings_activity, Toast.LENGTH_SHORT).show()
            return
        }
        startActivity(intent)
    }

    private fun openWatchLaterSettings() {
        val action = when {
            Build.VERSION.SDK_INT >= 31 -> Settings.ACTION_APP_OPEN_BY_DEFAULT_SETTINGS
            else -> Settings.ACTION_APPLICATION_DETAILS_SETTINGS
        }
        val intent = Intent(action, Uri.parse("package:${packageName}"))
        startActivity(intent)
    }

    companion object {
        private val EXAMPLE_URI = Uri.parse("https://www.youtube.com/watch?v=dGFSjKuJfrI")
        private val EXAMPLE_INTENT = Intent(Intent.ACTION_VIEW, EXAMPLE_URI)
    }
}
